dispatcher/dispatcher.go

```diff
@@ -30,122 +30,122 @@ func NewDispatcher(log *zap.SugaredLogger, queue queue.TaskQueue, db *db.DB) *Di
+
+	err := d.db.Attempts.BatchInsert(ctx, attempts)
 	if err != nil {
 		return err
 	}
 
-	for i, task := range tasks {
-		err := d.queue.Add(task, attempts[i].ScheduledAt.Time)
-		if err != nil {
-			d.log.Warnf("failed to add task to queue: %v", err)
-			continue
-		}
-		err = d.db.AttemptsWS.UpdateStatus(ctx, task.ID, entities.AttemptStatusQueued)
-		if err != nil {
-			d.log.Warnf("failed to update attempt status: %v", err)
-		}
-	}
+	d.pushTaksQueue(context.TODO(), attempts)
 
 	return nil
 }
 
-func (d *Dispatcher) dispatch(ctx context.Context, event *entities.Event, endpoints []*entities.Endpoint) error {
-	attempts := make([]*entities.Attempt, 0, len(endpoints))
-	tasks := make([]*queue.TaskMessage, 0, len(endpoints))
-
-	err := d.db.TX(ctx, func(ctx context.Context) error {
-		now := time.Now()
-		err := d.db.EventsWS.Insert(ctx, event)
-		if err != nil {
-			return err
-		}
-
-		for _, endpoint := range endpoints {
-			delay := endpoint.Retry.Config.Attempts[0]
-			attempt := &entities.Attempt{
-				ID:            utils.KSUID(),
-				EventId:       event.ID,
-				EndpointId:    endpoint.ID,
-				Status:        entities.AttemptStatusInit,
-				AttemptNumber: 1,
-				ScheduledAt:   types.NewTime(now.Add(time.Second * time.Duration(delay))),
-				TriggerMode:   entities.AttemptTriggerModeInitial,
-			}
-
-			task := &queue.TaskMessage{
-				ID: attempt.ID,
-				Data: &model.MessageData{
-					EventID:    event.ID,
-					EndpointId: endpoint.ID,
-					Attempt:    1,
-				},
-			}
-			attempts = append(attempts, attempt)
-			tasks = append(tasks, task)
+func (d *Dispatcher) pushTaksQueue(ctx context.Context, attempts []*entities.Attempt) {
+	for _, attempt := range attempts {
+		task := &taskqueue.TaskMessage{
+			ID: attempt.ID,
+			Data: &model.MessageData{
+				EventID:    attempt.EventId,
+				EndpointId: attempt.EndpointId,
+				Attempt:    attempt.AttemptNumber,
+			},
 		}
-
-		return d.db.AttemptsWS.BatchInsert(ctx, attempts)
-	})
-	if err != nil {
-		return err
-	}
-
-	for i, task := range tasks {
-		err := d.queue.Add(task, attempts[i].ScheduledAt.Time)
+		err := d.queue.Add(ctx, task, attempt.ScheduledAt.Time)
 		if err != nil {
 			d.log.Warnf("failed to add task to queue: %v", err)
 			continue
 		}
-		err = d.db.AttemptsWS.UpdateStatus(ctx, task.ID, entities.AttemptStatusQueued)
+		err = d.db.Attempts.UpdateStatus(ctx, task.ID, entities.AttemptStatusQueued)
 		if err != nil {
 			d.log.Warnf("failed to update attempt status: %v", err)
 		}
 	}
-
-	return nil
 }
 
-func listSubscribedEndpoints(ctx context.Context, db *db.DB, eventType string) (list []*entities.Endpoint, err error) {
+func (d *Dispatcher) listSubscribedEndpoints(ctx context.Context, wid, eventType string) (list []*entities.Endpoint, err error) {
 	var q query.EndpointQuery
-	endpoints, err := db.EndpointsWS.List(ctx, &q)
+	q.WorkspaceId = &wid
+	endpoints, err := d.db.Endpoints.List(ctx, &q)
```

Comment:
This function queries from db every time. Is it difficult to build an event routing layer?

Reply:
I feel this can be done in another PR. This PR is getting larger.